python/scripts/generate_sources.py

```diff
@@ -0,0 +1,414 @@
+            expression_clause = f'''\
+
+            if isinstance({argnames[0]}, Expression):
+                return Expression._call('{name}', {argstring})
+            '''
+        else:
+            expression_clause = ""
+
+        function_text = dedent(f'''\
+        def {name}{full_signature}:
+            """\n{funcdoc}
+            """
+            func = pyarrow._compute.get_function('{cpp_name}')            
+            {expression_clause}
+            return(
+                func.call({argstring}, memory_pool=memory_pool)
+            )
+        ''')
+    else:
+        # here we need to create the kwargs param substring for hte _handle_options function
+        if arity is Ellipsis or len(argnames) <= arity:
+            option_args = ''
+        else:
+            option_args = ', '.join(argnames[arity:]) + ','
+
+        if len(argnames) > 2 or (len(argnames) and not var_argname):
+            expression_clause = f'''\
+                
+            if isinstance({argnames[0]}, Expression):
+                return Expression._call('{name}', {argstring}, options)
+            '''
+        else:
+            expression_clause = ""
+
+        
+        function_text = dedent(f'''\
+        def {name}{full_signature}:
+            """\n{funcdoc}
+            """
+
+            options = _handle_options('{name}', {options_class_name}, options,
+                                    ({option_args}), {', '.join(opt + '=' + opt for opt in option_params)})
+            func = pyarrow._compute.get_function('{cpp_name}')
+            {expression_clause}
+            return(
+                func.call( {argstring}, options, memory_pool)
+            )
+        ''')
+
+    return(function_text)
+
+
+def write_compute_file(output_path):
+    """
+    Write the full set of generated functions to the output_path.
+    
+    Note that, in practice, while this generates core functions for all of the Arrow C++ compute functions, the compute functions may be overriden in the `pyarrow/compute.py` file. 
+
+    Args:
+        output_path: The full path to the file to write the compute functions to.
+    """
+    g = globals()
+    reg = function_registry()
+
+    doc_overrides = arrowdoc.parse_directory(os.path.dirname(arrowdoc.__file__) + "/../docs/additions/compute")
+
+    function_defs = []
+    for cpp_name in reg.list_functions():
+        name = function_name_rewrites.get(cpp_name, cpp_name)
+        
+        func = reg.get_function(cpp_name)
+        
+        if func.kind == "hash_aggregate":
+            # Hash aggregate functions are not callable,
+            # so let's not expose them at module level.
+            continue
+        
+        function_def = generate_function_def(name, cpp_name, func, func.arity, doc_overrides.get(name, None))
+        function_defs.append(function_def)
+    
+    with open(output_path, 'w') as fh:
+        fh.write(dedent(f"""\
+            import pyarrow
```

Comment:
We need an header here, something that mentions this is an autogenerated file and to not modify it.